xayn-ai/src/coi/system.rs

```diff
@@ -866,101 +960,203 @@ mod tests {
     }
 
     #[test]
-    fn test_compute_weights_empty_cois() {
+    fn test_compute_relevances_empty_cois() {
+        let config = Configuration {
+            horizon: Duration::from_secs_f32(SECONDS_PER_DAY),
+            ..Configuration::default()
+        };
+        let mut system = CoiSystem::new(config, SMBert::default());
         let cois = create_pos_cois(&[[]]);
-        let horizon = Duration::from_secs_f32(SECONDS_PER_DAY);
-        let weights = CoiSystem::default().compute_weights(&cois, horizon);
-        assert!(weights.is_empty());
+        system.compute_relevances(&cois);
+        assert!(system.coi_to_relevance.is_empty());
+        assert!(system.relevance_to_key_phrase.is_empty());
     }
 
     #[test]
-    fn test_compute_weights_zero_horizon() {
+    fn test_compute_relevances_zero_horizon() {
+        let config = Configuration {
+            horizon: Duration::ZERO,
+            ..Configuration::default()
+        };
+        let mut system = CoiSystem::new(config, SMBert::default());
         let cois = create_pos_cois(&[[1., 2., 3.], [4., 5., 6.]]);
-        let horizon = Duration::ZERO;
-        let weights = CoiSystem::default().compute_weights(&cois, horizon);
-        assert_approx_eq!(f32, weights, [0., 0.]);
+        system.compute_relevances(&cois);
+        assert_eq!(system.coi_to_relevance.len(), 2);
+        assert_approx_eq!(f32, system.coi_to_relevance[&cois[0].id], [0.]);
+        assert_approx_eq!(f32, system.coi_to_relevance[&cois[1].id], [0.]);
+        assert!(system.relevance_to_key_phrase.is_empty());
+    }
+
+    fn dedup<const N: usize>(array: [f32; N]) -> Array1<f32> {
+        assert!(IntoIterator::into_iter(array).all(f32::is_finite));
+        IntoIterator::into_iter(array)
+            .map(Relevance)
+            .collect::<BTreeSet<_>>()
+            .into_iter()
+            .map(Into::into)
+            .collect()
     }
 
     #[test]
-    fn test_compute_weights_count() {
+    fn test_compute_relevances_count() {
+        let config = Configuration {
+            horizon: Duration::from_secs_f32(SECONDS_PER_DAY),
+            ..Configuration::default()
+        };
+        let mut system = CoiSystem::new(config, SMBert::default());
         let mut cois = create_pos_cois(&[[1., 2., 3.], [4., 5., 6.], [7., 8., 9.]]);
         cois[1].stats.view_count += 1;
         cois[2].stats.view_count += 2;
-        let horizon = Duration::from_secs_f32(SECONDS_PER_DAY);
-        let weights = CoiSystem::default().compute_weights(&cois, horizon);
-        assert_approx_eq!(f32, weights, [0.5, 0.6666667, 0.8333333], epsilon = 0.00001);
+        system.compute_relevances(&cois);
+        assert_eq!(system.coi_to_relevance.len(), 3);
+        assert_approx_eq!(
+            f32,
+            system.coi_to_relevance[&cois[0].id],
+            0.5 * dedup(system.config.penalty),
+            epsilon = 0.00001,
```

Comment:
Seems that in the rest of the code `ulps` is used more, can we used that instead?

Reply:
this depends a lot on what we want to test here. the `ulps` is a measure for a relative error whereas the `epsilon` is a measure for an absolute error. for example both of the following will pass even though the absolute difference in the second case is huge:
```rust
assert_approx_eq!(f32, 9999999e-7, 9999998e-7, epsilon = 0., ulps = 2);
assert_approx_eq!(f32, 9999999e+10, 9999998e+10, epsilon = 0., ulps = 2);
```
in our case we know that the error is roughly about `1e-5` to `1-e6` due to the nature of the compared values and we can express this with the `epsilon` to have a more precise test.

Reply:
ok make sense. Thanks!